### ISSUE TITLE
Performance issues while zooming in

### ISSUE THREAD

**Issue by x6pnda** [googlemaps/js-markerclusterer#152]
Heyy @jpoehnelt,

I have been experiencing performance issues with the new version of the marker clusterer (1.0.17). I have a map with about 3500 markers and before we've been using an old version of markerclustererplus. I don't know the version as the version used is a .js file. The old clusterer was configured with the following settings:

`     
const markerCluster = new MarkerClusterer(map, [], {
            maxZoom: 15,
            minimumClusterSize: 5,
});`
        
This gives a fairly smooth experience. Now we upgraded to the new clusterer and any algorithm has a pretty big performance impact. We're using the following settings:

`
const markerClusterer = new MarkerClusterer({
                map: experienceMap,
                algorithm: new SuperClusterAlgorithm({ maxZoom: 15, minPoints: 5, radius: 60 }),
});`
            
There are no other event listeners on the map or on the markers. No custom icons or anything. Just a marker with a position and they're added to the map. Loading the initial makers is no problem and only takes about 15ms on an AMD 5900x with RTX 3070ti and 64gb ram. The problem begins when zooming in. It seems there is something blocking the main thread and as a result it all slows down and basically freezes. We've noticed this problem happening on Chrome, Edge and Firefox. We also tried the [demo site](https://googlemaps.github.io/js-markerclusterer/public/defaults/) and we experience the same performance problems to a lesser extent. After debugging we found out it causes in the reset() function of the MarkerClusterer class on the line with `this.clusters.forEach((cluster) => cluster.delete())`. The algorithm takes 8ms with 3500 markers but that line takes over 1700ms. It seems the clusters are getting rendered while the animation of the map zooming in is happening. 

Thanks. 

Update:
After testing it seems to be caused by the label tag of the marker. If all labels are removed, the lag is gone.

**Comment by x6pnda** [comment]
I have made a quick [video ](https://gyazo.com/1bd2a9c1598d7aed5e7a9bdf4760ad79) of whats happening. The blockiness is the real experience and not because of the video/gif. We have experience this on multiple systems with different OS installs.

**Comment by oshhh** [comment]
I'm facing a huge lag without labels as well. I want to plot 27k markers. @googlemaps/markerclustererplus worked fine for this scale.

**Comment by jpoehnelt** [comment]
What I see in the video is the expected behavior of adding and removing markers because every time the zoom changes the clusters change too. To minimize this you could customize the `Algorithm.calculate` function to only cluster when the zoom changes by some threshold (e.g. 5 levels) or calculate the same clusters for all zooms above some threshold and not cluster below that level. See https://github.com/googlemaps/js-markerclusterer/blob/main/src/algorithms/supercluster.ts.

**Comment by jpoehnelt** [comment]
some pseudocode

```ts
class ThresholdSuperClusterAlgorithm extends SuperClusterAlgorithm {
  private previousBucket: number = -1;

  public calculate(input: AlgorithmInput): AlgorithmOutput {
    const bucket = zoomBucket(input.map.getZoom(), 3);

    if (bucket === this.previousBucket) {
       this.previousBucket = bucket;
       return { clusters: this.clusters, changed: false };
    }
    this.previousBucket = bucket;

    return super.calculate(input);
}

const zoomBucket(zoom: number, factor: number): number => Math.floor(zoom / factor)
```

**Comment by x6pnda** [comment]
We already tried it. In the end we completely wrote our own algorithm and renderer which is faster than any of the ones provided but still not as fast as the old liberary. 

We put the old markerclusterer code next this code and the performance without labels is slightly worse than the old library. Exact same calculations, algorithms and renderer with same properties.  We also tried the new supercluster algorithm which was even slower compared to the old one. But oh well it might approve in some years, we wrote our own code. 

Now comes the fun part for you and the google maps teams. If you add labels to markers on a normal google maps map (no clusterer), the performance decreases massively even with a small amount of labels. 100 markers with labels is enough to have lag on a 16 core desktop. Then we tried the old clusterer with 4k markers which made about 100 clusters with labels inside of the markers. This doesn't create lag for some reason. Then we tried the new code and it causes more lag than the native google maps. 

In the end it's okay for me as it gave me a reason to work on a custom solution for my final project to get my bachelor but i'll definitely follow up when I have more time with the same application worked out in all 3 liberaries as my colleagues en me couldn't figure out what the problem is.
